`pulumi new` for Go projects doesn't fully prepare the project's dependencies

Running `pulumi new azure-go` initializes the new Go project and installs dependencies:

```
This command will walk you through creating a new Pulumi project.

Enter a value or leave blank to accept the (default), and press <ENTER>.
Press ^C at any time to quit.

project name: (azurego) 
project description: (A minimal Azure Native Go Pulumi program) 
Created project 'azurego'

Please enter your desired stack name.
To create a stack in an organization, use the format <org-name>/<stack-name> (e.g. `acmecorp/dev`).
stack name: (dev) 
Created stack 'dev'

azure-native:location: The Azure location to use: (WestUS) 
Saved config

Installing dependencies...

Finished installing dependencies

Your new project is ready to go! ✨

To perform an initial deployment, run 'pulumi up'
```

Subsequently running `pulumi up` yields the following errors:

```
Diagnostics:
  pulumi:pulumi:Stack (azurego-dev):
    main.go:4:2: missing go.sum entry for module providing package github.com/pulumi/pulumi-azure-native/sdk/go/azure/resources (imported by azurego); to add:
    	go get azurego
    main.go:5:2: missing go.sum entry for module providing package github.com/pulumi/pulumi-azure-native/sdk/go/azure/storage (imported by azurego); to add:
    	go get azurego
    main.go:6:2: missing go.sum entry for module providing package github.com/pulumi/pulumi/sdk/v3/go/pulumi (imported by azurego); to add:
    	go get azurego
 
    error: an unhandled error occurred: program exited with non-zero exit code: 1
```

Forcing you to run `go get $PROJECT` before `pulumi up` will work.

This isn't the intention. The "Installing dependencies..." step should take care of initializing the project so that `pulumi up` will work without requiring any other commands to be run first.

For Go projects, dependencies are initialized by `pulumi new` using `go mod download`. That used to be sufficient, but it looks like that's no longer the case with more recent versions of Go.

I think we can address this issue by changing `pulumi new` to use `go mod tidy` instead of `go mod download`.

I found the following in release notes: https://go.dev/blog/go116-module-changes

```
In Go 1.16, module-aware commands report an error after discovering a problem in go.mod or go.sum instead of attempting to fix the problem automatically. In most cases, the error message recommends a command to fix the problem.

$ go build
example.go:3:8: no required module provides package golang.org/x/net/html; to add it:
    go get golang.org/x/net/html
$ go get golang.org/x/net/html
$ go build
As before, the go command may use the vendor directory if it’s present (see Vendoring for details). Commands like go get and go mod tidy still modify go.mod and go.sum, since their main purpose is to manage dependencies.
```

It sounds like as of 1.16 `go mod download` will shy away from editing `go.mod` and `go.sum`. 

I'd like to retract my objection, as you point out this pertains to expanding templates, whether ours or user-defined templates, in which case `go mod tidy` sounds very appropriate - let `pulumi new` fixup `go.mod` and `go.sum`, this is fine.. This is not part of `pulumi up` and would not be affecting a stack that is in prod and tries to be conservative, just new templates. Let's take your fix.

--- a/CHANGELOG_PENDING.md
+++ b/CHANGELOG_PENDING.md
@@ -3,4 +3,5 @@
 
 ### Bug Fixes
 
-
+- [cli] - Avoid `missing go.sum entry for module` for new Go projects.
+  [#7808](https://github.com/pulumi/pulumi/pull/7808)

--- a/pkg/cmd/pulumi/new.go
+++ b/pkg/cmd/pulumi/new.go
@@ -652,7 +652,7 @@ func dotnetInstallDependenciesAndBuild(proj *workspace.Project, root string) err
 	return nil
 }
 
-// goInstallDependencies will install dependencies for the project by running `go mod download`
+// goInstallDependencies will install dependencies for the project by running `go mod tidy`.
 func goInstallDependencies() error {
 	fmt.Println("Installing dependencies...")
 	fmt.Println()
@@ -666,12 +666,12 @@ func goInstallDependencies() error {
 		return err
 	}
 
-	cmd := exec.Command(gobin, "mod", "download")
+	cmd := exec.Command(gobin, "mod", "tidy")
 	cmd.Env = os.Environ()
 	cmd.Stdout, cmd.Stderr = os.Stdout, os.Stderr
 
 	if err := cmd.Run(); err != nil {
-		return errors.Wrapf(err, "`go mod download` failed to install dependencies; rerun manually to try again, "+
+		return errors.Wrapf(err, "`go mod tidy` failed to install dependencies; rerun manually to try again, "+
 			"then run 'pulumi up' to perform an initial deployment")
 	}
 

